File: c++/078.cpp
class Solution {
/*
1. 结果集 res 初始化：添加一个空集
2. 向 res 的每一个元素加入 nums[i]，再添加至 res 中（原res不动，添加新元素）
*/
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        res.push_back({});
        for(auto x:nums){
            int len=res.size();
            for(int i=0; i<len; i++){
                vector<int> t=res[i];
                t.push_back(x);
                res.push_back(t);
            }
        }
        return res;
    }
};

File: go/078.go
package main

/**
子集
求数组所有的子集 不重复
经典回溯
循环递归，注意遍历元素 以及结果退回
*/
var r [][]int

func subsets(nums []int) [][]int {
	r = make([][]int, 0)
	path := make([]int, 0)
	f(nums, 0, path)
	return r
}

func f(nums []int, idx int, path []int) {
	t := make([]int, len(path))
	copy(t, path)
	r = append(r, t)
	for i := idx; i < len(nums); i++ {
		path = append(path, nums[i])
		idx++
		f(nums, idx, path)
		path = path[:len(path)-1]
	}
}
